File: swea/swea_1249.java
// Solve 2024-04-05

// SWEA에 제출할 때는 class 이름을 Solution으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;

public class swea_1249 {

    static class Point {
        int x, y, dist;

        Point(int x, int y, int dist) {
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
    }

    static int[] dx = { 0, 0, -1, 1 };
    static int[] dy = { -1, 1, 0, 0 };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int t = Integer.parseInt(br.readLine());

        for (int ti = 1; ti <= t; ti++) {
            int n = Integer.parseInt(br.readLine());
            int[][] board = new int[n][n];

            for (int x = 0; x < n; x++) {
                String line = br.readLine();

                for (int y = 0; y < n; y++) {
                    board[x][y] = line.charAt(y) - '0';
                }
            }

            int[][] minDist = new int[n][n];

            for (int x = 0; x < n; x++) {
                for (int y = 0; y < n; y++) {
                    minDist[x][y] = 1_000_000;
                }
            }

            PriorityQueue<Point> pqMinTop = new PriorityQueue<Point>(new Comparator<Point>() {

                @Override
                public int compare(Point o1, Point o2) {
                    return o1.dist < o2.dist ? -1 : o1.dist > o2.dist ? 1 : 0;
                }

            });

            minDist[0][0] = 0;
            pqMinTop.offer(new Point(0, 0, 0));

            while (!pqMinTop.isEmpty()) {
                Point cur = pqMinTop.poll();
                // System.out.println(cur.x + " " + cur.y + " " + cur.dist);

                if (cur.dist > minDist[cur.x][cur.y]) {
                    continue;
                }

                for (int d = 0; d < 4; d++) {
                    int nx = cur.x + dx[d];
                    int ny = cur.y + dy[d];

                    if (nx >= 0 && nx < n && ny >= 0 && ny < n) {
                        int nDist = cur.dist + board[nx][ny];

                        if (minDist[nx][ny] > nDist) {
                            minDist[nx][ny] = nDist;
                            pqMinTop.offer(new Point(nx, ny, nDist));
                        }
                    }
                }
            }

            sb.append("#").append(ti).append(" ").append(minDist[n - 1][n - 1]).append("\n");
        }

        System.out.print(sb);
        br.close();
    }

}

File: swea/swea_1249.cpp
// Solve 2024-01-14

#include <iostream>
#include <string>
#include <queue>

using namespace std;

struct Node{
    int x, y, d;

    Node(int x, int y, int d) : x(x), y(y), d(d) {}

    bool operator<(const Node &rhs) const {
        if (d != rhs.d) return d > rhs.d;
        if (x != rhs.x) return x > rhs.x;
        return y > rhs.y;
    }
};

const int INF = 1e9;
int dx[4] = { 0, 0, -1, 1 };
int dy[4] = { -1, 1, 0, 0 };
int board[100][100];
int dist[100][100];

int main(int argc, char** argv) {
    int t;
    cin >> t;

    for (int ti = 1; ti <= t; ti++) {
        int n;
        cin >> n;

        for (int i = 0; i < n; i++) {
            string s;
            cin >> s;

            for (int j = 0; j < n; j++) {
                board[i][j] = s[j] - '0';
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dist[i][j] = INF;
            }
        }

        dist[0][0] = 0;
        priority_queue<Node> pq_mintop;
        pq_mintop.emplace(0, 0, 0);

        while (!pq_mintop.empty()) {
            Node cur = pq_mintop.top();
            pq_mintop.pop();

            if (cur.x == n - 1 && cur.y == n - 1) break;
            if (cur.d > dist[cur.x][cur.y]) continue;

            for (int d = 0; d < 4; d++) {
                int nx = cur.x + dx[d];
                int ny = cur.y + dy[d];

                if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;

                int nd = cur.d + board[nx][ny];

                if (nd < dist[nx][ny]) {
                    dist[nx][ny] = nd;
                    pq_mintop.emplace(nx, ny, nd);
                }
            }
        }

        cout << '#' << ti << ' ' << dist[n - 1][n - 1] << '\n';
    }

    return 0;
}
